ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,256 @@
+        :param brain_name: The name of the brain associated with trainer config
+        :param reward_buff_cap: Max reward history to track in the reward buffer
+        :param trainer_parameters: The parameters for the trainer (dictionary).
+        :param training: Whether the trainer is set for training.
+        :param run_id: The identifier of the current run
+        """
+
+        super(GhostTrainer, self).__init__(
+            brain_name, trainer_parameters, training, run_id, reward_buff_cap
+        )
+
+        self.trainer = trainer
+
+        self.internal_policy_queues: List[AgentManagerQueue[Policy]] = []
+        self.internal_trajectory_queues: List[AgentManagerQueue[Trajectory]] = []
+
+        # assign ghost's stats collection to wrapped trainer's
+        self.stats_reporter = self.trainer.stats_reporter
+
+        self_play_parameters = trainer_parameters["ghost"]
+        self.window = self_play_parameters["window"]
+        self.current_prob = self_play_parameters["current_prob"]
+        self.steps_between_snapshots = self_play_parameters["snapshot_per"]
+
+        self.policies: Dict[str, TFPolicy] = {}
+        self.policy_snapshots: List[Any] = [None for _ in range(self.window)]
+        self.snapshot_counter: int = 0
+        self.learning_behavior_name: str = None
+        self.current_policy_snapshot = None
+        self.last_step = 0
+
+        self.initial_elo: float = 1200.0
+        self.current_elo: float = self.initial_elo
+        self.policy_elos: List[float] = [self.initial_elo for _ in range(self.window)]
+        self.current_opponent: int = 0
+
+    @property
+    def get_step(self) -> int:
+        """
+        Returns the number of steps the trainer has performed
+        :return: the step count of the trainer
+        """
+        return self.trainer.get_step
+
+    def _write_summary(self, step: int) -> None:
+        """
+        Saves training statistics to Tensorboard.
+        """
+        opponents = np.array(self.policy_elos)
+        LOGGER.info(
+            " ELO: {}\n"
+            "Mean Opponent ELO: {}"
+            " Std Opponent ELO: {}".format(
+                self.current_elo, opponents.mean(), opponents.std()
+            )
+        )
+        self.stats_reporter.add_stat("ELO", self.current_elo)
+
+    def _process_trajectory(self, trajectory: Trajectory) -> None:
+        if (
+            trajectory.done_reached
+            and not trajectory.max_step_reached
+            and self.current_opponent > -1
+        ):
+            result = "win"
+            final_reward = trajectory.steps[-1].reward
+            if final_reward < 0:
+                result = "loss"
+            change = compute_elo_rating_changes(
+                self.current_elo, self.policy_elos[self.current_opponent], result
+            )
+            self.current_elo += change
+            self.policy_elos[self.current_opponent] -= change
+
+    def _is_ready_update(self) -> bool:
+        pass
```

Comment:
return False? Probably equivalent but it's not doing what the type hint says.

Reply:
True enough. Fixed.  It's never actually called since I override the advance() function.